Carousel Improvements

tobi-or-not-tobi opened the pull request:
Draft for all the changes of #9481

Commit:
merge dev

tobi-or-not-tobi committed:
Merge branch 'develop' into carousel

tobi-or-not-tobi committed:
start lazy load items

tobi-or-not-tobi committed:
add prefetching

tobi-or-not-tobi committed:
add ghosts

tobi-or-not-tobi committed:
fix prefetch

tobi-or-not-tobi committed:
accessible carousel items beyond prefetching

tobi-or-not-tobi committed:
temp fixes

tobi-or-not-tobi committed:
fix latest selected carousel indicator

tobi-or-not-tobi committed:
Merge branch 'develop' into carousel

tobi-or-not-tobi committed:
Merge branch 'develop' into carousel

tobi-or-not-tobi committed:
loads of improvements, code starts to become cleaner

tobi-or-not-tobi committed:
small improvements

tobi-or-not-tobi committed:
fix

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-9481

tobi-or-not-tobi committed:
start adding product images

LTiger14 commented on projects/storefrontlib/src/shared/components/carousel/carousel.component.ts:
```diff
@@ -62,32 +82,204 @@ export class CarouselComponent implements OnInit {
+   */
+  protected freeze() {
+    this._freeze = true;
+    window.setTimeout(() => {
```
Please use windowRef

LTiger14 commented on projects/storefrontlib/src/shared/components/carousel/carousel.component.ts:
```diff
@@ -38,11 +44,25 @@ export class CarouselComponent implements OnInit {
+   */
+  @Input() focusConfig: FocusConfig = {
+    autofocus: true,
+    // focusOnScroll: true,
```
Cleanup

LTiger14 commented on projects/storefrontlib/src/layout/loading/intersect.directive.ts:
```diff
@@ -0,0 +1,56 @@
+  ) {}
+
+  protected get config(): IntersectionOptions {
+    // TODO: read the default config from the intersectionService->LayoutConfig
```
You could introduce optional `threshold` to the config is you want to keep backwards compatibility.

LTiger14 commented on projects/storefrontlib/src/layout/a11y/keyboard-focus/lock/lock-focus.directive.ts:
```diff
@@ -158,7 +158,7 @@ export class LockFocusDirective
   handleEscape(event: KeyboardEvent): void {
     if (this.shouldLock) {
-      this.service.clear(this.config.group);
+      // this.service.clear(this.config.group);
```
Why did you change this?

LTiger14 commented on projects/storefrontlib/src/cms-components/product/carousel/product-carousel/product-carousel.component.html:
```diff
@@ -1,19 +1,40 @@
+<!-- *ngIf="(title$ | async) === 'Our Bestselling Products'" -->
```
Cleanup

LTiger14 commented on projects/storefrontlib/src/shared/components/carousel/carousel.model.ts:
```diff
@@ -0,0 +1,27 @@
+import { ICON_TYPE } from '../../../cms-components/misc/icon/icon.model';
+
+export interface CarouselButton {
+  // visible: boolean;
```
I don't think we need the visible property.

tobi-or-not-tobi committed:
add initial carousel ghost set

tobi-or-not-tobi committed:
fix ff and safari visually

LTiger14 commented on projects/storefrontlib/src/shared/components/carousel/carousel.component.html:
```diff
@@ -1,61 +1,87 @@
+  <div
+    #carousel
+    class="carousel-panel"
+    [ngClass]="'size-' + size"
```
Isn't the `size-...` class logic commented out of the css?

tobi-or-not-tobi replied:
yes, it's gone now.

LTiger14 commented on projects/storefrontlib/src/layout/a11y/keyboard-focus/tab/tab-focus.service.ts:
```diff
@@ -13,18 +13,19 @@ export class TabFocusService extends AutoFocusService {
           ? this.findNextScrollable(host, config, increment)
           : this.findNext(host, config, increment);
 
+      console.log('next', next);
```
Cleanup

LTiger14 commented on projects/storefrontlib/src/layout/a11y/keyboard-focus/autofocus/auto-focus.directive.ts:
```diff
@@ -50,6 +54,22 @@ export class AutoFocusDirective
     }
+
+    if (this.config.focusOnScroll) {
+      // TODO: cleanup
```
Cleanup

LTiger14 commented on projects/storefrontlib/src/shared/components/carousel/carousel.component.html:
```diff
@@ -1,61 +1,87 @@
+            *ngFor="let data of items | slice: i:i + size; let itemNum = index"
+            class="item"
+            [class.active]="i === activeSlide"
+            (mouseenter)="prefetch()"
```
Is this mouseenter required? I find it executes quite a bit of logic every time you hover an item. Obviously it won't send new network call when the data is already fetched but moving your mouse back and forth leads to a lot of function calls.

tobi-or-not-tobi replied:
It is, but the current separate data we pass into the carousel item template is causing loads of change detection cycles. I believe we must change this into a single object with observables for the item, prefetch and tabindex. I need to test it out though.

LTiger14 commented on projects/storefrontstyles/scss/components/product/carousel/_carousel.scss:
```diff
@@ -1,118 +1,288 @@
 %cx-carousel {
```
CSS cleanup general comment 😉

LTiger14 reviewed (changes requested):
In my opinion, the direction the carousel is heading into is very good. That said, there are a few areas I would like to improve.

1. The styling in Firefox and Safari needs a bit of tweaking. I found that despite the odd looks, the behaviour was still correct on both browsers. Additionally, the carousel didn't really like my Iphone 11 Pro simulator.

Firefox 82:
<img width="1252" alt="Screen Shot 2020-10-23 at 1 42 44 PM" src="https://user-images.githubusercontent.com/10605724/97197711-568af000-1784-11eb-976f-6301dff173d3.png">

Safari:
<img width="963" alt="Screen Shot 2020-10-23 at 1 42 34 PM" src="https://user-images.githubusercontent.com/10605724/97197776-67d3fc80-1784-11eb-9f56-3f46e951300f.png">
<img width="943" alt="Screen Shot 2020-10-23 at 1 42 38 PM" src="https://user-images.githubusercontent.com/10605724/97197797-6dc9dd80-1784-11eb-8ce0-54f7788f0e9d.png">

IOS safari, Iphone 11 PRO:
<img width="543" alt="Screen Shot 2020-10-26 at 12 01 28 PM" src="https://user-images.githubusercontent.com/10605724/97197847-7d492680-1784-11eb-86a0-398707da70b2.png">

2. I did some logging and I noticed that the `getProduct(code: string, prefetch: boolean)` in `product-carousel.component` got called a TON of times. I understand that there is some prefetching going on, I added a comment regarding the `mouseover` on carousel items that added a few calls.

3. The carousel has always been some pretty complicated code to understand but I feel like now it's gotten a step more difficult. I'm sorry I don't have a good suggestion to simplify it at the moment; perhaps if we get together we could find something.

**For after**: I added some cleanup comments so we don't forget.

saad-mo commented:
Hello - @tobi-or-not-tobi @LTiger14, This PR contains a11y fixes for keyboard focus which would be helpful to have in place while we address Screen Reader issues. What is the likely hood of this getting done anytime soon?

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-9481

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-9481

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-9481

tobi-or-not-tobi committed:
merge to unblock stashing

tobi-or-not-tobi committed:
heavy refactor

- improved accessibility; you can tab through items and previous/next buttons, the focus moves to the next logical item when using keyboard
- start to embrace directionality, some work left on product images thumb scenario
- more refactor is needed for cleaner/more logical code

tobi-or-not-tobi committed:
Merge branch 'develop' into feature/GH-9481

tobi-or-not-tobi committed:
Move references to carousel out of the core component lib

tobi-or-not-tobi committed:
split up carousel modules in sep libs
